src/analysis/processing/qgsalgorithmfileuploader.cpp

```diff
@@ -0,0 +1,156 @@
+}
+
+QgsFileUploaderAlgorithm *QgsFileUploaderAlgorithm::createInstance() const
+{
+  return new QgsFileUploaderAlgorithm();
+}
+
+void QgsFileUploaderAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFile( QStringLiteral( "FILE" ), QObject::tr( "File to upload" ), Qgis::ProcessingFileParameterBehavior::File, QString(), QVariant(), false, QObject::tr( "All file (%1)" ).arg( QLatin1String( "*.*" ) ) ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "URL" ), tr( "To URL" ), QVariant(), false, false ) );
+
+  auto formNameParam = std::make_unique<QgsProcessingParameterString>( QStringLiteral( "FORMNAME" ), tr( "Form name field" ), QString(), false, true );
+  formNameParam->setHelp( QObject::tr( "The optional form name field parameter emulate a filled-in form in which a user has pressed the submit button. This enables uploading of binary files when url end point require a form name key" ) );
+  formNameParam->setFlags( formNameParam->flags() | Qgis::ProcessingParameterFlag::Optional );
+  addParameter( formNameParam.release() );
+}
+
+QVariantMap QgsFileUploaderAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mFeedback = feedback;
+  QString url = parameterAsString( parameters, QStringLiteral( "URL" ), context );
+  if ( url.isEmpty() )
+    throw QgsProcessingException( tr( "No URL specified" ) );
+
+  const QString filePath = parameterAsFile( parameters, QStringLiteral( "FILE" ), context );
+  const bool exists = QFileInfo::exists( filePath );
+  if ( !feedback->isCanceled() && !exists )
+    throw QgsProcessingException( tr( "The file %1 doesn't exist." ).arg( filePath ) );
+
+  const QString formNameKey = parameterAsString( parameters, QStringLiteral( "FORMNAME" ), context );
+  QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT" ), context );
+
+  QEventLoop loop;
+  QTimer progressTimer;
+  QUrl uploadUrl;
+  QStringList errors;
+
+
+  QgsFileUploader *uploader = new QgsFileUploader( filePath, url, formNameKey );
+
+  connect( mFeedback, &QgsFeedback::canceled, uploader, &QgsFileUploader::cancelUpload );
+  connect( uploader, &QgsFileUploader::uploadError, this, [&errors, &loop]( const QStringList &e ) { errors = e; loop.exit(); } );
+  connect( uploader, &QgsFileUploader::uploadProgress, this, &QgsFileUploaderAlgorithm::receiveProgressFromUploader );
+  connect( uploader, &QgsFileUploader::uploadCompleted, this, [&uploadUrl]( const QUrl url ) { uploadUrl = url; } );
+  connect( uploader, &QgsFileUploader::uploadExited, this, [&loop]() { loop.exit(); } );
+  connect( &progressTimer, &QTimer::timeout, this, &QgsFileUploaderAlgorithm::sendProgressFeedback );
+  uploader->startUpload();
+
+  progressTimer.start( 1000 );
+
+  loop.exec();
```

Comment:
this isn't thread safe -- if the algorithm is run on the main thread (which can happen), then triggering an event loop will cause issues.

Can you adopt the approach used by QgsBlockingNetworkRequest (or possibly just add the upload API to that class!) to avoid this?